src/shared/tools.h

```diff
@@ -567,11 +593,105 @@ static inline bool htcmp(GLuint x, GLuint y)
 
 template <class T> struct vector
 {
+private:
     static const int MINSIZE = 8;
 
     T *buf;
     int alen, ulen;
+public:
+    //some std::vector like interfaces
+    void insert(T* position, T const* start, T const* end )
+    {
+        insert( position - buf, start, end - start );
+    }
+
+    void resize( int sz )
+    {
+        reserve( sz - ulen );
+        setsize( sz );
+    }
+
+    T* data( void )
+    {
+        return buf;
+    }
+
+    T const* data( void ) const
+    {
+        return buf;
+    }
+
+    void clear( void )
+    {
+        setsize( 0 );
+    }
+
+    T * begin( void )
+    {
+        return buf;
+    }
+
+    T * end( void )
+    {
+        return buf + length();
+    }
+
+    T const* begin( void ) const
+    {
+        return buf;
+    }
+
+    T const* end( void ) const
+    {
+        return buf + length();
+    }
+
+    int size( void ) const
+    {
+        return length();
+    }
+
+    void erase( T const* it )
+    {
+        remove( it - buf );
+    }
 
+    void erase( T const* start, T const* end )
+    {
+        remove( start - buf, end - buf );
+    }
+
+    void pop_back( void )
+    {
+        pop();
+    }
+
+    void emplace_back( T && el )
+    {
+        add( el );
```

Comment:
Does `add` make use of the rvalue reference, i.e., not copy but move the item?

Reply:
not really. Yes, I could have called this push_back, but since my long term intent is to allow removing the nih vector implementations, I only added emplace_back for now.

Reply:
Ok, now I remember why I choose emplace_back: it allows the use of add(), and yes, I shamelessly duplicated the code and change it for the various variants.